### ISSUE TITLE
Turning on accessible canvas settings causes sketches to break

### ISSUE THREAD

**Issue by davepagurek** [processing/p5.js-web-editor#2188]
#### Details about the bug:

- p5.js version: 1.6.0
- Web browser and version: Firefox 111.0.1 (also reported from another user on Chrome)
- Operating System: MacOS 12.5.1
- Steps to reproduce this:

Example sketch:
```js
function setup() {
  createCanvas(400, 400);
}

function draw() {
  background('red');
}
```
 
Then check this box in settings:
![image](https://user-images.githubusercontent.com/5315059/228248733-0d4e1ad5-15fd-4c7c-bdc3-ad91cde72201.png)

While normally this would just make a red canvas, with that setting turned on, we just get an error:
![image](https://user-images.githubusercontent.com/5315059/228248890-67b95384-7b5a-4080-909e-2bd393347a16.png)

The error is coming from the p5.accessibility library editing the behaviour of colour functions:
![image](https://user-images.githubusercontent.com/5315059/228249361-8bf1ab79-72d8-4650-9f9f-f3c51b4ce09f.png)

Looking at the repo here https://github.com/processing/p5.accessibility it seems like it's no longer maintained. Is there an alternative that can be used? In the mean time, it seems to be broken.

**Comment by lindapaiste** [comment]
Good catch!  It's a real bummer that this causes the whole sketch not to render.  

> Looking at the repo here https://github.com/processing/p5.accessibility it seems like it's no longer maintained. Is there an alternative that can be used? In the mean time, it seems to be broken.

I started writing out a bunch of stuff about wrapping the `p5.accessibility` library to catch errors, but then I realized that the accessible output functionality has been merged into the core `p5.js` package for quite a while.

We should be using the accessible output from `p5.js` rather than adding the `p5.accessibility` script.  So I need to figure out how to configure that.  We potentially need to check the p5 version and show a "friendly error" if the user has accessible outputs turned on but the p5 version in their code is not high enough.

<details>

<summary>
what I started writing
</summary>

I wonder if it's possibly for us to put some sort of wrapper around the `p5.accessibility` function calls with a `try`/`catch` to prevent them from breaking the canvas output.  This would go in the "previewApp" code.  The `p5.accessibility` code runs automatically based on the presence of certain HTML ids, but all of its variables are available in the global/`window` scope of the preview iframe.

I need to test this out, but it's something like:
```
if (window.baseInterceptor && window.baseInterceptor.prototype) {
  const originalFn = window.baseInterceptor.prototype.getColorName;
  window.baseInterceptor.prototype.getColorName = function(colArgs) {
    try {
      return originalFn(colArgs);
    } catch (e) {
      // or in this case, we do know the color 'red' and it's in colArgs[0]
      return `Unknown color `${colArgs.join(', ')}`;
    }
  }
}
```
Or maybe a broader wrapper, like around all functions on `window.baseInterceptor.prototype`.

The hard part would be making sure that our replacement executes before the package executes.

I'm so confused by this code because it looks like they're checking for a "does not match"  `!`  when it should be a "does match"?
https://github.com/processing/p5.accessibility/blob/165b836a8153aeb0529a678a211efac112f3d4fe/src/baseInterceptor.js#L88-L90
</details>

**Comment by davepagurek** [comment]
I'm not super familiar with the accessibility features of p5, is this https://p5js.org/reference/#/p5/textOutput sort of what the accessibility script was doing before? Would we need to inject `textOutput()` or `gridOutput()` into scripts when these web editor settings are enabled?

**Comment by lindapaiste** [comment]
> I'm not super familiar with the accessibility features of p5, is this https://p5js.org/reference/#/p5/textOutput sort of what the accessibility script was doing before? Would we need to inject textOutput() or gridOutput() into scripts when these web editor settings are enabled?

Me neither!  We need to inject those functions but I need to figure out where and how.  I think it only needs to be called one time in `setup()` rather than on every `draw()`?

If that's the right move then we would need to A) add a line in the `setup()` if that function already exists B) add a `setup()` function if there is not one in the script already.  Messing with JS files as text strings could be fragile as it's easy to mess up if the user writes code abnormally.  I don't know if it's more durable to parse it as an Abstract Syntax Tree using `jscodeshift`?  The easiest thing would be to not modify the user's code at all, and instead just add an additional script file with some global call but I'm not sure if that works?  Possibly works when using p5 globally like most users of the editor do.  Almost certainly would not work with instance mode.

Here's how the `p5.accessibility` package works:
1. Global script runs once when the `.js` file is loaded.
2. It [defines a global variable `allData`](https://github.com/processing/p5.accessibility/blob/master/src/_data.js) which is a huge (688KB) JSON object containing schemas for all of the p5 classes and methods that it should "spy" on.  This was last updated in 2017!
3. It looks for DOM elements with ids [`'textOutput-content'`](https://github.com/processing/p5.accessibility/blob/165b836a8153aeb0529a678a211efac112f3d4fe/src/textInterceptor/interceptorP5.js#L24) and/or [`'tableOutput-content'`](https://github.com/processing/p5.accessibility/blob/165b836a8153aeb0529a678a211efac112f3d4fe/src/gridInterceptor/interceptorP5.js#L23) and uses these as containers to render its output.  If neither is found then it doesn't do anything.
4. It [wraps all of the spied-on methods on the global `p5.prototype` object](https://github.com/processing/p5.accessibility/blob/165b836a8153aeb0529a678a211efac112f3d4fe/src/gridInterceptor/interceptorP5.js#L35).  Each method is replaced with a version that generates and renders the accessible output each time that it is called.

Based on what I'm seeing, I doubt that it does anything when using p5 in instance mode?

The current web editor handling for accessible outputs is here:
https://github.com/processing/p5.js-web-editor/blob/2cd9067d4395f29d9176da7bf781b49fd0851d2c/client/modules/Preview/EmbedFrame.jsx#L227-L248

It's all in the `index.html` file of the project rather than the `.js`.

If the `textOutput` or `gridOutput` options are enabled, then it injects a `<script src="https://cdn.jsdelivr.net/gh/processing/p5.accessibility@0.1.1/dist/p5-accessibility.js">`  tag into the document and also injects the necessary output containers `<section id="textOutput-content">` or `<section id="tableOutput-content">`.

**Comment by lindapaiste** [comment]
This is somewhat tangential, but I noticed that we have a property in the redux state, `state.ide.isAccessibleOutputPlaying` which does not do anything at the moment.

https://github.com/processing/p5.js-web-editor/blob/269ca2292886b76a474278b87f35376ad1c0bd62/client/modules/IDE/pages/IDEView.jsx#L368-L373
We have this very perplexing line in the `IDEView` which is just a `div` with a `boolean` in it.  It does not do anything, and renders to the DOM as and empty `div`.

I had to go way back in the history to figure out what this was *supposed* to be doing.  The `boolean` value was previously used to conditionally render an `AccessibleOutput` component.  [Since this change in 2018](https://github.com/processing/p5.js-web-editor/commit/24b0be629fcf58c8a649db4e7b8a25fa1cbc02a1#diff-253e5906b9c0d59e1a03121da2b6f87d0cbebd6a85dac2dffce5bbc29f9aefcc) (PR #508) it does not do anything.

**Where it's set**
There is a [second version of the Play button](https://github.com/processing/p5.js-web-editor/blob/dde4940ad193aa88e0367489d95cecb45011b234/client/modules/IDE/components/Toolbar.jsx#L97-L120) which is hidden off-screen and intended to be seen by screen readers.  When this button is pressed, it dispatches a `startAccessibleSketch` action instead of the normal `startSketch` action.  That action sets the `isAccessibleOutputPlaying` flag in the state to `true`.   Clicking that button also dispatches `setTextOutput(true)` and `setGridOutput(true)`, so we *are* enabling the accessible output (by setting the `textOutput` and `gridOutput` preferences which are normally set through the Preferences modal). Therefore the `isAccessibleOutputPlaying` flag is redundant and unnecessary.

**Comment by davepagurek** [comment]
> If that's the right move then we would need to A) add a line in the setup() if that function already exists B) add a setup() function if there is not one in the script already.

Another option which might be easier than parsing the source code could be to make something that works like a p5 plugin and [register a method to run when the sketch is initialized.](https://github.com/processing/p5.js/blob/main/contributor_docs/creating_libraries.md#use-registermethod-to-register-functions-with-p5-that-should-be-called-at-various-times) p5 has an [internal `_setup` that is always called](https://github.com/processing/p5.js/blob/14e9b51e8863b906984cbd9964f505be0e6b0fc9/src/core/main.js#L324) regardless of whether there's a proper `setup` function, so we could maybe augment the existing `_setup` to call `textOutput` if it hasn't already been called:

```js
p5.prototype.ensureAccessibleCanvas = function() {
  const prevSetup = this._setup;
  this._setup = () => {
    prevSetup();

    // Call textOutput if it has not already been called
    if (!this._accessibleOutputs.text) {
      this.textOutput();
    }
  };
};
p5.prototype.registerMethod('init', p5.prototype.ensureAccessibleCanvas);
```

**Comment by lindapaiste** [comment]
@davepagurek That sounds much smarter!  Thanks for your insight on this.  I haven't used `registerMethod` before so I wasn't sure how this could be done on a global level.

**Comment by davepagurek** [comment]
btw one more thought that crossed my mind: it might be less brittle to not override `_setup` and just use the pre-draw hook, since the `this._accessibleOutputs.text)` check would already prevent it from double-calling:

```js
p5.prototype.ensureAccessibleCanvas = function() 
  // Call textOutput if it has not already been called
  if (!this._accessibleOutputs.text) {
    this.textOutput();
  }
};
p5.prototype.registerMethod('pre', p5.prototype.ensureAccessibleCanvas);
```
